### DISCUSSION TITLE
syntax when processing DICOM input files

### DISCUSSION THREAD

**Discussion by zeffiro** [muscbridge/PyDKE#304] [Technical Support] [answered]
When input files are converted to NIFTI format for processing the different NIFTI files are listed, separated by commas. When DICOM format files that are spread across multiple directories are the inputs, should the directories be put in a comma-separated list as well?

**Comment by TheJaeger** [comment]
@zeffiro that is correct, different directories of sorted DICOM files can be entered together. The latest version of PyDesigner does away with comma-separated files though - you'd specify them with a space now.

**Reply by zeffiro** [reply]
Thank you for the prompt response. I will give it a try.

**Reply by zeffiro** [reply]
I was wondering if you could help me getting NeuroDocker running with some dMRI data collected with the ABCD MRI protocol.

Here is the directory structure:
E:/BEAP-dMRI/sub-BEAP001/ses-1/
│
├── dwi
│ ├── ABCD_DMRI_0015 (directory containing DICOM files)
│ ├── ABCD_DMRI_DISTORTIONMAP_AP_0017 (directory containing DICOM files for TOPUP)
│ ├── 
└── derivative/dwi (empty dir)

Here is the call I tried:
docker run -it --rm -v E:/BEAP-dMRI/sub-BEAP001/ses-1/dwi:/data \
dmri/neurodock pydesigner --standard\
--output E:/BEAP-dMRI/sub-BEAP001/ses-1/derivative/dwi\
ABCD_DMRI_0015  ABCD_DMRI_DISTORTIONMAP_AP_0017

And here is the error I am getting:
Traceback (most recent call last):
  File "/usr/local/bin/pydesigner", line 10, in <module>
    sys.exit(main())
  File "/usr/local/lib/python3.7/dist-packages/designer/pydesigner.py", line 262, in main
    image = DWIParser(args.dwi)
  File "/usr/local/lib/python3.7/dist-packages/designer/preprocessing/util.py", line 364, in __init__
    ftype.append(mrinfoutil.format(self.DWIlist[i]).lower())
  File "/usr/local/lib/python3.7/dist-packages/designer/preprocessing/mrinfoutil.py", line 61, in format
    type = getconsole(path, '-format')
  File "/usr/local/lib/python3.7/dist-packages/designer/preprocessing/mrinfoutil.py", line 32, in getconsole
    raise OSError('Input path does not exist. Please ensure that the '
OSError: Input path does not exist. Please ensure that the folder or file specified exists.

Am I making a syntax error in the command line call?

**Reply by TheJaeger** [reply, marked as answer]
@zeffiro Yea, your files are not specified relative to the mount point you've created. The contents of the directory `E:/BEAP-dMRI/sub-BEAP001/ses-1/dwi` will be available to the Docker container under `/data/*`. Which means your output directory `E:/BEAP-dMRI/sub-BEAP001/ses-1/derivative/dwi` is not visible to the container, and also needs to be specified relative to `/data`. Not only that, the input path needs to be an absolute path, which I presume is mounted at `/data/ABCD_DMRI_0015 ABCD_DMRI_DISTORTIONMAP_AP_0017 `.

**Comment by zeffiro** [comment]
Getting the mount points correct solved the problem. It is running now. Many thanks.

What does PyDesigner do if the command line contains distortion scans with both AP and PA encoding?
This is the ABCD default,e.g.  ABCD_DMRI_DISTORTIONMAP_PA_0014  ABCD_DMRI_DISTORTIONMAP_AP_0015

**Reply by zeffiro** [reply]
Please ignore the previous question. I discovered that both PE images (AP and PA) need to be included for topup to run.

Getting that far uncovered another problem: eddy_cuda does not see to run, so it defaults to the CPU version:

...
Command:  eddy_cuda10.2 --imain=eddy_in.nii --mask=eddy_mask.nii --acqp=eddy_config.txt --index=eddy_indices.txt --bvecs=bvecs --bvals=bvals --topup=field --repol --data_is_shelled --out=dwi_post_eddy --verbose

dwifslpreproc: CUDA version of 'eddy' was not successful; attempting OpenMP version

Command:  eddy_cpu --imain=eddy_in.nii --mask=eddy_mask.nii --acqp=eddy_config.txt --index=eddy_indices.txt --bvecs=bvecs --bvals=bvals --topup=field --repol --data_is_shelled --out=dwi_post_eddy --verbose
..

But, I have installed the NVIDIA driver outside on docker before and am able to run CUDA software, e.g.
conda install pytorch torchvision torchaudio pytorch-cuda=11.8 -c pytorch -c nvidia

Is there an additional step needed to install the CUDA drivers in the docker image?

**Reply by TheJaeger** [reply]
That's right @zeffiro, PyDesigner figures out all phase encoding directions in your dataset and applies the appropriate corrections.

CUDA will not work within Docker implementation of PyDesigner. However, there are workarounds. If your host system is Linux, then native installation of FSL, MRtrix3, PyDesigner, and CUDA drivers will get CUDA version of eddy current correction going.

It's a little trickier on Windows, but it's doable. You'd need Windows 11 and Windows Susbsystem for Linux (WSL2) for GPU passthrough. You'd need to install your preferred Linux environment on WSL2, then install FSL, MRtrix3, and PyDesigner into it. CUDA applications will work within WSL2 as long as you have the most recent Nvidia drivers and Windows 11.

**Comment by zeffiro** [comment]
Here is the associated error message:

eddy_cuda10.2: error while loading shared libraries: libcuda.so.1: cannot open shared object file: No such file or directory
